internal/handlers/common/update.go

```diff
@@ -18,18 +18,175 @@ import (
 	"fmt"
 	"sort"
 	"strings"
+	"time"
 
 	"github.com/FerretDB/FerretDB/internal/types"
 	"github.com/FerretDB/FerretDB/internal/util/must"
 )
 
+// CheckCurrentdateFieldExpression checks $currentDate input on correctness.
+func CheckCurrentdateFieldExpression(update *types.Document) error {
+	for _, updateOp := range update.Keys() {
+		if updateOp != "$currentDate" {
+			continue
+		}
+
+		currentDateExpression, err := update.Get("$currentDate")
+		if err != nil {
+			return nil
+		}
+
+		switch currentDateExpression := currentDateExpression.(type) {
+		case *types.Document:
+			for _, field := range currentDateExpression.Keys() {
+				setValue := must.NotFail(currentDateExpression.Get(field))
+
+				switch setValue := setValue.(type) {
+				case bool:
+					// ok
+
+				case *types.Document:
+					for _, k := range setValue.Keys() {
+						if k != "$type" {
+							return NewWriteErrorMsg(
+								ErrBadValue,
+								fmt.Sprintf("Unrecognized $currentDate option: %s", k),
+							)
+						}
+					}
+					currentDateType, err := setValue.Get("$type")
+					if err != nil { // default is date
+						continue
+					}
+
+					switch currentDateType := currentDateType.(type) {
+					case string:
+						switch currentDateType {
+						case "timestamp":
+							return NewWriteErrorMsg(
+								ErrNotImplemented,
+								"timestamp of $surrentDate is not implemented",
+							)
+						case "date":
+							// ok
+						default:
+							return NewWriteErrorMsg(
+								ErrBadValue,
+								"The '$type' string field is required to be 'date' or 'timestamp'",
+							)
+						}
+
+					default:
+						return NewWriteErrorMsg(
+							ErrBadValue,
+							"The '$type' string field is required to be 'date' or 'timestamp'",
+						)
+					}
+
+				default:
+					return NewWriteErrorMsg(
+						ErrBadValue,
+						fmt.Sprintf("%s is not valid type for $currentDate. Please use a boolean ('true') "+
+							"or a $type expression ({$type: 'timestamp/date'}).", AliasFromType(setValue),
+						),
+					)
+				}
+			}
+		default:
+			return NewWriteErrorMsg(
+				ErrFailedToParse,
+				"Modifiers operate on fields but we found another type instead",
+			)
+		}
+	}
```

Comment:
I see here 4 defaults here, and it's comfortable to compare with the original structure where data change applies. 
But maybe you have other suggestions. What are your thoughts?